core/src/nucleus/actions/call_zome_function.rs

```diff
@@ -0,0 +1,144 @@
+extern crate futures;
+use crate::{
+    action::{Action, ActionWrapper},
+    context::Context,
+    nucleus::{
+        is_fn_public,
+        ribosome::{self, api::call::check_capability},
+        ZomeFnCall, ZomeFnResult,
+    },
+};
+use futures::{
+    future::Future,
+    task::{LocalWaker, Poll},
+};
+use holochain_core_types::error::HolochainError;
+use std::{pin::Pin, sync::Arc};
+
+use holochain_core_types::json::JsonString;
+use std::thread;
+
+#[derive(Clone, Debug, PartialEq, Hash)]
+pub struct ExecuteZomeFnResponse {
+    call: ZomeFnCall,
+    result: ZomeFnResult,
+}
+
+impl ExecuteZomeFnResponse {
+    pub fn new(call: ZomeFnCall, result: Result<JsonString, HolochainError>) -> Self {
+        ExecuteZomeFnResponse { call, result }
+    }
+
+    /// read only access to call
+    pub fn call(&self) -> ZomeFnCall {
+        self.call.clone()
+    }
+
+    /// read only access to result
+    pub fn result(&self) -> Result<JsonString, HolochainError> {
+        self.result.clone()
+    }
+}
+
+/// Initialize Application, Action Creator
+/// This is the high-level initialization function that wraps the whole process of initializing an
+/// instance. It creates both InitApplication and ReturnInitializationResult actions asynchronously.
+///
+/// Returns a future that resolves to an Ok(NucleusStatus) or an Err(String) which carries either
+/// the Dna error or errors from the genesis callback.
+///
+/// Use futures::executor::block_on to wait for an initialized instance.
+pub async fn call_zome_function(
+    zome_call: ZomeFnCall,
+    context: &Arc<Context>,
+) -> Result<JsonString, HolochainError> {
+    let (dna_name, code) = {
+        let state = context.state().ok_or(HolochainError::ErrorGeneric(
+            "Context not initialized".to_string(),
+        ))?;
+        let nucleus_state = state.nucleus();
+        let dna = nucleus_state
+            .dna
+            .as_ref()
+            .ok_or(HolochainError::DnaMissing)?;
+
+        // 1. Validate the call (a number of things could go wrong)
+        // 1.a make sure the zome and function exists
+        let _ = dna
+            .get_function_with_zome_name(&zome_call.zome_name, &zome_call.fn_name)
+            .map_err(|e| HolochainError::Dna(e))?;
```

Comment:
This can be shorthanded to:
```suggestion
            .map_err(HolochainError::Dna)?;
```